File: sobol.py
import numpy as np
from scipy.stats import qmc
import matplotlib.pyplot as plt

# Sobol sequence generator helper class
class SobolN:
    def __init__(self, ndim, skip):
        self.ndim = ndim
        self.skip = skip
        self.sampler = qmc.Sobol(d=self.ndim, scramble=False)
        self.sampler.fast_forward(self.skip)

    def rndm_array(self):
        return self.sampler.random(1)[0]

# Main function
def main(npoints=5000, nskip=0):
    ndim = 2  # Dimension of the sequence
    sobol_generator = SobolN(ndim, nskip)

    # Generate pseudo-random numbers
    pseudo_random_points = np.random.rand(npoints, ndim)

    # Generate Sobol quasi-random numbers
    sobol_points = np.array([sobol_generator.rndm_array() for _ in range(npoints)])

    # Plotting
    fig, axs = plt.subplots(1, 2, figsize=(12, 6))

    # Pseudo-random plot
    axs[0].scatter(pseudo_random_points[:, 0], pseudo_random_points[:, 1], s=5, c="blue")
    axs[0].set_title("Pseudo-Random")
    axs[0].set_xlabel("x1")
    axs[0].set_ylabel("x2")
    axs[0].set_xlim(0, 1)
    axs[0].set_ylim(0, 1)
    axs[0].grid(True)

    # Sobol quasi-random plot
    axs[1].scatter(sobol_points[:, 0], sobol_points[:, 1], s=5, c="red")
    axs[1].set_title("Sobol (Quasi-Random)")
    axs[1].set_xlabel("x1")
    axs[1].set_ylabel("x2")
    axs[1].set_xlim(0, 1)
    axs[1].set_ylim(0, 1)
    axs[1].grid(True)

    # Save the plot as a PDF
    plt.tight_layout()
    plt.savefig("random.pdf")
    plt.show()

# Run the main function
if __name__ == "__main__":
    import sys

    # Parse command line arguments
    npoints = int(sys.argv[1]) if len(sys.argv) > 1 else 5000
    nskip = int(sys.argv[2]) if len(sys.argv) > 2 else 0

    main(npoints, nskip)
    

File: sobol.cpp
//+______________________________________________________________________________
// Example of generating quasi-random numbers

#include "TApplication.h"
#include "Math/QuasiRandom.h"
#include "TRandom3.h"
#include "TGraph.h"
#include "TH2F.h"
#include "TROOT.h"
#include "TCanvas.h"
#include "TGClient.h"
#include "TStyle.h"
#include <cstdlib>
#include <iostream>

using namespace std;
using namespace ROOT::Math;


// This helper class is used to instantiate the sobel sequence generator for a given dimension
// skip may be used to start at some offset from the beginning of the sequence

class SobolN {
public:
  SobolN(int ndim, unsigned long skip){
    r1=new QuasiRandomSobol(ndim);
    r1->Skip(skip);
  }
  void RndmArray(double *array) {
    r1->Next(array);
  }
private:
  QuasiRandomSobol *r1;
};



// run as
// ./sobol
// or to adjust the numebr of points and offset in the sobel sequence
// eg ./sobol 2000 10000

int main(int argc, char **argv) {

  unsigned long npoints=5000;
  unsigned long nskip=0;
  
  // This allows you to view ROOT-based graphics in your C++ program
  // If you don't want view graphics (eg just read/process/write data files), 
  // this can be ignored
  TApplication theApp("App", &argc, argv);

  if (argc>1) npoints=atol(argv[1]);
  if (argc>2) nskip=atol(argv[2]);
  
  gStyle->SetOptStat(0);

  TGraph *gr = new TGraph();
  TGraph *gs = new TGraph();

  TRandom3 *tr3 = new TRandom3(0);
  
  // quasi random numbers need to be created giving the dimension of the sequence
  // in this case we generate a 2-d sequence
  const int ndim=2;
  SobolN rsobel(ndim,nskip);

  TH2F *h = new TH2F("h","",10,0,1,10,-1,1);
   
  // generate n random number pairs
  double *x = new double[ndim];

  for (int i = 0; i < npoints; ++i){
    tr3->RndmArray(ndim,x);
    gr->SetPoint(i,x[0],x[1]);
    rsobel.RndmArray(x);
    gs->SetPoint(i,x[0],x[1]);
  }
   
  UInt_t dw = gClient->GetDisplayWidth();
  UInt_t dh = gClient->GetDisplayHeight();

  TCanvas * c1 = new TCanvas("c1","Random sequence",dw/2,dh/3);
  c1->Divide(2,1);


  auto hr=new TH2F("hr","PseudoRandom;x1;x2",10,0,1,10,0,1);
  auto hs=new TH2F("hs","Sobol;x1;x2",10,0,1,10,0,1);
  c1->cd(1);
  hr->Draw();
  gr->Draw("p");
  
  c1->cd(2);
  hs->Draw();
  gs->Draw("p");

  c1->SaveAs("random.pdf");
  
  // view graphics in ROOT if we are in an interactive session
  if (!gROOT->IsBatch()) {
    cout << "To exit, quit ROOT from the File menu of the plot (or use control-C)" << endl;
    theApp.SetIdleTimer(30,".q"); // set up a failsafe timer to end the program
    theApp.Run(true);
  }

  delete [] x;
  return 0;
}
